### DISCUSSION TITLE
Proper way to scale Valhalla

### DISCUSSION THREAD

**Discussion by newcarrotgames** [valhalla/valhalla#3902] [Q&A]
I'm using AWS to host Valhalla. Can anyone recommend the best instance type, storage, etc to host Valhalla with? For instance, is the app more compute or memory intensive? I've attached the results of a load test done using the m5xlarge instance type. Are these results typical? There seems to be a sharp increase in the average request time when dealing with multiple requests, so I'm thinking the best option may be to use several smallish instance types and a load balancer, but I wanted to get the community's opinion first. Maybe someone has already gone through this process and could give me some advice. Thank you for reading!

**Comment by kevinkreiser** [comment]
Can you first tell us a bit about how you're running and testing it? Which executables are you running, are you matching the number of threads/procs to the number of cores, what type of tile storage are you using, what types of requests are you testing with?

**Reply by newcarrotgames** [reply]
Flask and gunicorn run the valhalla service itself while the tiles are stored in S3 initially and then copied into each EC2 instance. We're mainly after the `/isochrone` endpoint, so most of the requests I'm using to test are a random lat/lon (within a range). Currently, there are 20 m5.4xlarge ec2 instances, but I feel that we may not need quite that much. Overall, CPU never seems to get above 25%, so I wonder if we'd be better off using memory-optimized instances.

**Reply by newcarrotgames** [reply]
@kevinkreiser Is that enough information, or were you expecting something else?

**Reply by kevinkreiser** [reply]
Are you storing the tiles in a tar file? How many contours and how large of an isochrone? These things all affect performance. It's not uncommon to spend relatively small cpu because most of the time is spent on i/o getting the tile data from the disk to ram and finally to the cpu. You should definitely use an nvme drive.

**Reply by newcarrotgames** [reply]
A single contour for 60 minutes by car takes around 3 seconds. Is that normal? The tiles are stored as a tar file.

It sounds like we should use storage optimized instances: https://aws.amazon.com/ec2/instance-types/i3/

Is there another format other than a basic tar file we should consider?

**Reply by kevinkreiser** [reply]
That does not sound normal, sounds too slow. Maybe for the first request but after the server is warmed up it shouldn't be that long. Have you tried it locally on your own machine just for a comparison? The tar file is the fastest option at the moment.

**Reply by newcarrotgames** [reply]
I tried rebuilding the tiles and running valhalla on an ec2 instance that uses nvme drives but I'm still seeing the >2.5s response times for 60 minutes searches. However, looking at the demo API, this seems normal:
```
https://valhalla1.openstreetmap.de/isochrone?json={"locations":[{"lat":40.677494, "lon":-73.904982}],"costing":"auto","contours":[{"time":60}],"date_time":{"type":1,"value":"2023-01-24T09:19"},"denoise":0.0,"generalize":100.0}
```
From my machine, that takes between 2.5s and 3s. Any ideas on how this search can be optimized? Maybe my request isn't structured the best way?

**Reply by newcarrotgames** [reply]
I've been testing other products (like graphhopper and openrouteservice) and similar searches for 60 minute isochrones seem to take ~30-60 seconds, so kudos to the Valhalla team.

**Reply by nilsnolde** [reply]
I also think 3 seconds is quite decent for a 60 mins car isochrone, at least for dense areas (the more edges, the longer it takes).

**Reply by kevinkreiser** [reply]
to me this seems slow :) iirc I think the latency should be less than half a second. 3 seconds is worrisome in my opinion. we really need like nightly performance tests and a graph we can watch over time correlated to merges to master.. then write don't have to guess what is changing or not changing performancewise

**Reply by newcarrotgames** [reply]
In some cases, the response time is less than 1s but this seems to, of course, depend on the complexity of the network surrounding the initial point. The point I've chosen for the above search is a random location in Brooklyn which I assumed would be slower than a rural area.

**Reply by nilsnolde** [reply]
Well, that’s expanding almost entire NYC (at least without traffic;)). I stand by 3 secs is fine, but maybe my expectations are too low, coming from an ORS background😄
